Added Durations

dominickj-tdi opened the pull request:
I've implemented the concept of Durations in this library, something I've found missing in my attempts to use it in a project.  This also provides some simple mathematical operations that can now be done on a Calendar.

dominickj-tdi committed:
Added CalendarDuration and related functionality
The goal here is to allow something to calculate distances between dates, do some basic addition and substraction, etc...

dominickj-tdi committed:
Made Calendar interface match changes to GregorianCalendar

dominickj-tdi committed:
Added operators to GregorianCalendar

dominickj-tdi committed:
Added tests for calculateDurationToDate()

dnfield commented on lib/src/calendar.dart:
```diff
@@ -52,4 +66,24 @@ abstract class Calendar implements Comparable<Calendar> {
+  /// If you subtract a [CalendarDuration] from this, the function
+  /// will return a new Calendar that is the given distance back
+  /// in time.
+  dynamic operator -(dynamic other);
```
This seems confusing.  I would rather just have named methods that are clearer about what you're adding or subtracting.

dominickj-tdi replied:
Fair enough. Although I prefer the more compact syntax of a simple operator, using only the named methods is also fine. `addCalendarDuration()` and `calculateDurationToDate()` implement this functionality so we can just delete the add and subtract operators.

dominickj-tdi replied:
I've added another commit which removes these operators.

dnfield commented on lib/src/gregorian_calendar.dart:
```diff
@@ -98,7 +59,7 @@ class GregorianCalendar implements Calendar {
   /// The number of days in the month of this date
   int get monthLength {
-    final int n = _daysPerMonth[_month];
+    final int n = daysPerMonth[_month];
```
This seems bad to make public.  It will be very easy to misuse.

dominickj-tdi replied:
I've only made it public to the rest of the library, so that gregorian_calendar.dart and gregorian_calendar_duration.dart could share constants. I intentionally did not add the new gregorian_consts.dart file I created to the export list for that very reason.

Do you have an alternate suggestion to allow sharing of constants between the two files?

dnfield replied:
Ahh I see now.  That's ok then.

dnfield commented on lib/src/gregorian_calendar.dart:
```diff
@@ -113,27 +74,90 @@ class GregorianCalendar implements Calendar {
 
     final int zeroBased =
-        (y + y ~/ 4 - y ~/ 100 + y ~/ 400 + _sakamotoHelper[month - 1] + day) %
+        (y + y ~/ 4 - y ~/ 100 + y ~/ 400 + sakamotoHelper[month - 1] + day) %
```
Ditto.

dominickj-tdi committed:
Remvoved + and - operators from Calendar.

dnfield commented on lib/src/gregorian_calendar_duration.dart:
```diff
@@ -0,0 +1,157 @@
+      days:  newReferenceDate.day - referenceDate.day
+    );
+  }
+
```
nit: additional whitespace here.

dnfield replied:
And elsewhere in this file.

dominickj-tdi replied:
Corrected. Whitespace should now be consistent.

dnfield commented on lib/src/gregorian_calendar_duration.dart:
```diff
@@ -0,0 +1,157 @@
+
+  @override
+  GregorianCalendarDuration operator -(){
+    return new GregorianCalendarDuration(
```
Why not normalize `-` as well?

dominickj-tdi replied:
My thought was that, if the user is inverting it, they will either have already normalized it, or they might for whatever reason not want it normalized.

dnfield commented on lib/src/gregorian_calendar_duration.dart:
```diff
@@ -0,0 +1,157 @@
+
+  @override
+  num toComparableNum(){
+    return (years * avgDaysPerYear) + (months * avgDaysPerMonth) + (weeks * daysPerWeek) + days;
```
This will be prone to precision errors, no?

dominickj-tdi replied:
Yes, there would be precision errors involved. I'm not sure if there is really an excellent way to accomplish comparisons since in the real world years and months don't divide nice and evenly into days. I feel like it's probably good enough for most purposes though. In situations requiring exact precision, the user should select a reference date and use `CalendarDuration.toDays(referenceDate)`. 

We could probably improve consistency a little by doing `(years * monthsPerYear + months) * avgDaysPerMonth`. That would at least guarantee that 1 year always exactly equals 12 months.

I guess the other option we could do is have a hard-coded reference date to perform comparisons against. I don't really like that idea because it seems like a lot of extra processing overhead and code complexity, plus doing that would lead to inconsistent-feeling behavior such as 1 month > 30 days but 2 months < 60 days (assuming the reference date was January 1st of a non-leap year)

dominickj-tdi replied:
Another thought would be to truncate the result to an integer and rename the method to `toDaysEstimated` to indicate it's a less precise version of `toDays`

dominickj-tdi replied:
Method removed in recent commits.

dnfield commented on lib/src/gregorian_calendar_duration.dart:
```diff
@@ -0,0 +1,157 @@
+
+  @override
+  int compareTo(CalendarDuration<GregorianCalendar> other) {
+    return toComparableNum().compareTo(other.toComparableNum());
```
This doesn't seem safe due to the precision errors that may arise in `toComparableNum`

dominickj-tdi replied:
Method removed in most recent commit

dominickj-tdi committed:
Switched toComparableNum to toApproxDays

This rounds to an integer to ensure that prescision errors don't cause equal time periods to be marked unequal.

dominickj-tdi committed:
Updated tests to match changes in previous commit

dnfield commented on lib/src/gregorian_calendar_duration.dart:
```diff
@@ -0,0 +1,147 @@
+
+  @override
+  int toApproxDays(){
+    return ((years * monthsPerYear + months) * avgDaysPerMonth).round() + (weeks * daysPerWeek) + days;
```
Why can't we just use a similar formula to what already existed?

Basically get a normalization of the date, and then get an int that is more or less ISO-8601 without the dashes?

dominickj-tdi replied:
The problem I see with that would be in situations where the combined weeks and days exceed the length of a month or year. We can't definitively normalize days into years or months without a reference date, since years and months have a variable number of days. If we try to normalize 35 days, it could be 1 month 4 days, 1 month 5 days, 1 month 6 days, or 1 month 7 days. 

With the current normalization function, 35 days normalizes as 5 weeks. if we were to do what it sounds like you're saying, (that is, concatenate years, months, weeks, and days into a single integer) then 35 days would be something like 500, whereas 1 month would be something like 10000, meaning 35 days < 1 month. Obviously that would be wrong.

The advantage to using something like the `toApproxDays()` is that it is totally agnostic to a reference date, and still accurate to what one would typically expect. 1 year == 365 days, 1 month == 30 days, etc.

dnfield replied:
Ahh right. I think that's a fundamental problem with adding durations then - this library is meant to make it easier to use calendar dates correctly.  Maybe the duration needs to be pegged to a year so we know how to count the days properly for the months?

dnfield replied:
(Introducing approximations like this is bound to cause trouble for people who don't realize it's only approximate, and it can cause trouble in ways that are very hard to reproduce)

dominickj-tdi replied:
You're right, this is a fundamental problem with the nature of the system. However, a duration is a very important real-world concept that applies in countless situations. I do feel like this is something that is needed.

I guess the only real solution is to disallow directly comparing durations to one another. Force people to either explicitly call `toDays(referenceDate)` or `toApproxDays()`

dominickj-tdi replied:
I have changed the == operator to only mark durations as equal if they are for sure exactly equal. No more approximations or averages. I have removed all other comparison operators. Now the only way durations can be compared is by explicitly converting to an integer via `toApproxDays()` (so there can be no question to the user this is an approximation) or by calling `toDays(referenceDate)` so an exact comparison can be made compared to a specific date.

dominickj-tdi committed:
Removed > and < operators from Duration.
This is becuase behavior for comparisons can be difficult to define in an objective and 100% consistent way.
Use `toApproxDays()` or `toDays(referenceDate)` to perform comparisons.

dominickj-tdi committed:
Fixed hash code
after the most recent change, two Durations might be marked equal but have different hash codes. This is now fixed.

dominickj-tdi commented:
I get the impression you would prefer to avoid having durations in this library. As it appears our opinions differ on the subject, I will close this pull request and continue with my own fork of the project. 

Thank you for your comments as they have certainly helped me improve my code. It's been a pleasure working with you.

dnfield commented:
Sorry - I've been busy with a few other things.  

I'm still open to this, I'd just like to avoid any approximations in this library.  Please feel free to reopen this or a new one if you'd like to continue towards that, otherwise you're more than welcome to continue with your fork.

dominickj-tdi commented:
Sure thing. I'll continue working independently for the time being. I feel like, where now approximations have to be explicitly invoked by the user, and are never invoked internally, there should be little chance of it becoming a problem. Thanks again for your suggestions however, they have been helpful!
